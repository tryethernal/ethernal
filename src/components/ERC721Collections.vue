<template>
    <v-container fluid>
        <div class="d-flex align-center mb-4">
            <h2 class="text-h6 font-weight-medium flex-grow-1">All NFT Collections</h2>
        </div>
        <v-divider class="my-4"></v-divider>
        <v-card>
            <v-card-text>
                <v-data-table-server
                    class="hide-table-count"
                    :loading="loading"
                    :items="tokens"
                    :items-length="tokenCount"
                    :headers="headers"
                    :sort-by="[{ key: currentOptions.orderBy, order: currentOptions.order }]"
                    :must-sort="true"
                    :sort-desc="true"
                    items-per-page-text="Rows per page:"
                    no-data-text="No ERC-721 collections found"
                    last-icon=""
                    first-icon=""
                    :items-per-page-options="[
                        { value: 10, title: '10' },
                        { value: 25, title: '25' },
                        { value: 100, title: '100' }
                    ]"
                    item-key="address"
                    @update:options="getTokens">
                    <template v-slot:item.address="{ item }">
                        <Hash-Link :type="'token'" :hash="item.address" :contract="item" />
                    </template>
                    <template v-slot:item.tokenName="{ item }">
                        {{ item.tokenName }}
                    </template>
                    <template v-slot:item.tokenSymbol="{ item }">
                        {{ item.tokenSymbol }}
                    </template>
                    <template v-slot:item.tokenTotalSupply="{ item }">
                        {{ item.tokenTotalSupply ? parseInt(item.tokenTotalSupply).toLocaleString() : 'N/A' }}
                    </template>
                    <template v-slot:item.tags="{ item }">
                        <v-chip v-for="(pattern, idx) in item.patterns" :key="idx" size="x-small" color="success" class="mr-2">
                            {{ formatContractPattern(pattern) }}
                        </v-chip>
                    </template>
                </v-data-table-server>
            </v-card-text>
        </v-card>
    </v-container>
</template>

<script setup>
import { ref, onMounted, onUnmounted } from 'vue';
import HashLink from '@/components/HashLink.vue';
import { formatContractPattern } from '@/lib/utils';
import { inject } from 'vue';

// Inject dependencies
const $pusher = inject('$pusher');
const $server = inject('$server');

// Reactive state
const loading = ref(true);
const tokens = ref([]);
const tokenCount = ref(0);
const currentOptions = ref({ 
    page: 1, 
    itemsPerPage: 10, 
    orderBy: 'timestamp', 
    order: 'desc', 
    pattern: 'erc721' 
});

// Table headers
const headers = [
    { title: 'Address', key: 'address' },
    { title: 'Name', key: 'tokenName' },
    { title: 'Symbol', key: 'tokenSymbol' },
    { title: 'Total Supply', key: 'tokenTotalSupply' },
    { title: '', key: 'tags', sortable: false }
];

// Methods
const getTokens = async ({ page, itemsPerPage, sortBy } = {}) => {
    loading.value = true;

    if (!page || !itemsPerPage || !sortBy || !sortBy.length) {
        loading.value = false;
        return;
    }

    if (currentOptions.value.page === page && 
        currentOptions.value.itemsPerPage === itemsPerPage && 
        currentOptions.value.sortBy === sortBy[0].key && 
        currentOptions.value.sort === sortBy[0].order) {
        loading.value = false;
        return;
    }

    const options = {
        page,
        itemsPerPage,
        orderBy: sortBy[0].key,
        order: sortBy[0].order,
        pattern: 'erc721'
    };

    try {
        const { data } = await $server.getContracts(options);
        tokens.value = data.items;
        tokenCount.value = data.items.length === currentOptions.value.itemsPerPage
            ? (currentOptions.value.page * data.items.length) + 1
            : currentOptions.value.page * data.items.length;
    } catch (error) {
        console.log(error);
    } finally {
        loading.value = false;
    }
};

// Lifecycle hooks
let newNftPusherHandler = null;
let destroyedContractPusherHandler = null;

onMounted(() => {
    newNftPusherHandler = $pusher.onNewNft(() => getTokens(currentOptions.value));
    destroyedContractPusherHandler = $pusher.onDestroyedContract(() => getTokens(currentOptions.value));
});

onUnmounted(() => {
    if (newNftPusherHandler) {
        newNftPusherHandler();
    }
    if (destroyedContractPusherHandler) {
        destroyedContractPusherHandler.unbind(null, null);
    }
});
</script>
