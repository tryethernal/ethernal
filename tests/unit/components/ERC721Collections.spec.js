import MockHelper from '../MockHelper';
import flushPromises from 'flush-promises'

import ERC721Collections from '@/components/ERC721Collections.vue';

const helper = new MockHelper();

describe('ERC721Collections.vue', () => {
    beforeEach(() => jest.clearAllMocks());

    it('Should load & display erc721 contracts', async () => {
        jest.spyOn(helper.mocks.server, 'getContracts')
            .mockResolvedValue({ data: { total: 1, items: [{ address: '0x123', tokenName: 'Ethernal', tokenSymbol: 'ETL', tokenTotalSupply: 100, patterns: ['erc721'] }]}});
        
        const wrapper = helper.mountFn(ERC721Collections, {
            stubs: ['Hash-Link']
        });
        await flushPromises();
        
        expect(wrapper.html()).toMatchSnapshot();
    });
});
