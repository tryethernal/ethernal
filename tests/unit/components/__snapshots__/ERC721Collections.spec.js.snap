// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ERC721Collections.vue Should load & display erc721 contracts 1`] = `
<div class="container container--fluid">
  <div class="v-data-table v-data-table--has-bottom theme--light">
    <div class="v-data-table__wrapper">
      <table>
        <colgroup>
          <col class="">
          <col class="">
          <col class="">
          <col class="">
          <col class="">
        </colgroup>
        <thead class="v-data-table-header">
          <tr>
            <th role="columnheader" scope="col" aria-label="Address: Not sorted. Activate to sort ascending." aria-sort="none" class="text-start sortable"><span>Address</span><i aria-hidden="true" class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light" style="font-size: 18px;"></i></th>
            <th role="columnheader" scope="col" aria-label="Name: Not sorted. Activate to sort ascending." aria-sort="none" class="text-start sortable"><span>Name</span><i aria-hidden="true" class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light" style="font-size: 18px;"></i></th>
            <th role="columnheader" scope="col" aria-label="Symbol: Not sorted. Activate to sort ascending." aria-sort="none" class="text-start sortable"><span>Symbol</span><i aria-hidden="true" class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light" style="font-size: 18px;"></i></th>
            <th role="columnheader" scope="col" aria-label="Total Supply: Not sorted. Activate to sort ascending." aria-sort="none" class="text-start sortable"><span>Total Supply</span><i aria-hidden="true" class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light" style="font-size: 18px;"></i></th>
            <th role="columnheader" scope="col" aria-label="Not sorted. Activate to sort ascending." aria-sort="none" class="text-start sortable"><span></span><i aria-hidden="true" class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light" style="font-size: 18px;"></i></th>
          </tr>
        </thead>
        <tbody>
          <tr class="">
            <td class="text-start">
              <hash-link-stub hash="0x123" type="nft"></hash-link-stub>
            </td>
            <td class="text-start">
              Ethernal
            </td>
            <td class="text-start">
              ETL
            </td>
            <td class="text-start">
              100
            </td>
            <td class="text-start"><span class="success mr-2 v-chip v-chip--no-color theme--light v-size--x-small"><span class="v-chip__content">
                ERC 721
            </span></span></td>
          </tr>
        </tbody>
      </table>
    </div>
    <div class="v-data-footer">
      <div class="v-data-footer__select">Rows per page:<div class="v-input v-input--hide-details v-input--is-label-active v-input--is-dirty theme--light v-text-field v-select">
          <div class="v-input__control">
            <div role="button" aria-haspopup="listbox" aria-expanded="false" aria-owns="list-16" class="v-input__slot">
              <div class="v-select__slot">
                <div class="v-select__selections">
                  <div class="v-select__selection v-select__selection--comma">10</div><input aria-label="Rows per page:" id="input-16" readonly="readonly" type="text" aria-readonly="false" autocomplete="off">
                </div>
                <div class="v-input__append-inner">
                  <div class="v-input__icon v-input__icon--append"><i aria-hidden="true" class="v-icon notranslate mdi mdi-menu-down theme--light"></i></div>
                </div><input type="hidden" value="10">
              </div>
              <div class="v-menu">
                <!---->
              </div>
            </div>
          </div>
        </div>
      </div>
      <div class="v-data-footer__pagination">1-1 of 1</div>
      <div class="v-data-footer__icons-before"><button type="button" disabled="disabled" class="v-btn v-btn--disabled v-btn--icon v-btn--round v-btn--text theme--light v-size--default" aria-label="Previous page"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate mdi mdi-chevron-left theme--light"></i></span></button></div>
      <div class="v-data-footer__icons-after"><button type="button" disabled="disabled" class="v-btn v-btn--disabled v-btn--icon v-btn--round v-btn--text theme--light v-size--default" aria-label="Next page"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate mdi mdi-chevron-right theme--light"></i></span></button></div>
    </div>
  </div>
</div>
`;
